### ISSUE TITLE
New Dask Arrow-based strings cause test failures

### ISSUE THREAD

**Issue by jakirkham** [dask/dask-image#335]
(Edited by @m-albert)

In the presence of `pyarrow`, dask by default assumes dataframes of type object to be pyarrow strings (see https://github.com/dask/dask/issues/10139#issuecomment-1655928619).

This creates problems revealed by failing tests (e.g. `test_dask_image/test_ndmeasure/test_find_objects.py::test_3d_find_objects`)

https://github.com/dask/dask-image/blob/67540af25597f84e4a642d644ba30dce7aebe753/dask_image/ndmeasure/_utils/_find_objects.py#L68-L70

`dd.from_delayed(df1, meta=meta).compute().dtypes`

Working install:
> 0    object
1    object
2    object
dtype: object

Failing install:
>0    string[pyarrow]
1    string[pyarrow]
2    string[pyarrow]
dtype: object

The failing test had come up when releasing v2023.08.0 in https://github.com/conda-forge/dask-image-feedstock/pull/14.

@jakirkham found that `pyarrow` is installed with the [conda distribution of dask](https://github.com/conda-forge/dask-feedstock/blob/89ff1ea9116e0798d9073d819a6b9b22fa1d0917/recipe/meta.yaml#L31), but not when [installing over pip](https://github.com/dask/dask/blob/main/pyproject.toml), where it just part of the `[complete]` target.

Also @jakirkham found that the above described conflicting behaviour can be [turned off using the dask configuration](https://github.com/dask-contrib/dask-sql/commit/af180620216793b7c6125b7db8f98dd0bba038de#diff-a31c7ed5d35f5ed8233994868c54d625b18e6bacb6794344c4531e62bd9dde59R15-R16).

He did this for the tests performed by the dask-image conda feedstock on [v2023.08.0](https://github.com/conda-forge/dask-image-feedstock/pull/14).

**Comment by jakirkham** [comment]
Please fill free to edit and fill this issue out Marvin 🙂

Just needed a placeholder for tracking

**Comment by jakirkham** [comment]
Would also be good to make a note in issue (where feedback is being collected): https://github.com/dask/dask/issues/10139

Ideally with a simple reproducer

**Comment by m-albert** [comment]
Despite the passing tests, potentially users who installed dask-image over conda would still experience the above described problem when using `ndmeasure.find_objects` (need to check whether there's more affected).

Perhaps a suitable fix for this on the `dask-image` side would be to set `dask.config.set({"dataframe.convert-string": False})` using a context manager around the affected functionality?

**Comment by GenevieveBuckley** [comment]
> Would also be good to make a note in issue (where feedback is being collected): [dask/dask#10139](https://github.com/dask/dask/issues/10139)
> 
> Ideally with a simple reproducer

I've made [a comment here](https://github.com/dask/dask/issues/10139#issuecomment-1664877956), but no reproducer (I'm not planning to do more work on this, it's open for anyone who wants it)

**Comment by jakirkham** [comment]
Yeah think we are not seeing this in CI as it requires a newer version of Dask than we are testing. Perhaps we should upgrade one of the CI environments (like 3.11) to a very recent Dask version

Tbh I've not looked deeply into the Dask Arrow work. Have heard about it mainly in passing. So not sure how `dask.config` handles it

Should add this pain point is not unique to us. We had to disable this feature in Dask-SQL recently as well ( https://github.com/dask-contrib/dask-sql/pull/1206 ). Unclear whether this is due to upstream bugs or if we need to make changes

**Comment by GenevieveBuckley** [comment]
> Yeah think we are not seeing this in CI as it requires a newer version of Dask than we are testing. Perhaps we should upgrade one of the CI environments (like 3.11) to a very recent Dask version

We could add an "upstream" CI environment, that just uses whatever the latest (or even pre-release?) versions are, maybe?

**Comment by jakirkham** [comment]
There are Dask nightly packages. So that would be easy to add

**Comment by m-albert** [comment]
As far as I understand, for reproducing the test failure in CI, next to a recent dask version we'd need `arrow>=7` present in the environment. This is a regular dependency of the conda , but not the pypi dask distribution.

**Comment by jakirkham** [comment]
Is this still an issue with recent Dask releases? Asking as they may have fixed something upstream since this occurred

**Comment by GenevieveBuckley** [comment]
I'm [not seeing any flaky/failing tests](https://github.com/dask/dask-image/actions/runs/8058913780), so I don't think this is still happening currently. I'll close the issue, and if it pops up again we can re-open.

**Comment by jakirkham** [comment]
Yeah I think part of the issue before was CI doesn't capture this edge case. Though maybe it should

**Comment by m-albert** [comment]
Reopening as this issue just came up in https://github.com/dask/dask-image/issues/355.

**Comment by jakirkham** [comment]
Think this may have been fixed in the intervening time. The test suite no longer fails for me

**Comment by m-albert** [comment]
This problem now occurs not only with dask installed over conda.

As described above, the problem occurs in the context of using `pyarrow` based strings in `dask.dataframe`. In `dask>=2025.2.0`, the implementation of `dask.dataframe` has been completely replaced by a different one previously living in `dask-expr` and since `2025.1.0` `pyarrow` is a direct dependency of `dask.dataframe` (see https://github.com/dask/dask/commit/a817ce951cfe30d5edbcea89eafa4583e2617f24).

With the new dask.dataframe implementation, the workaround tested [above](https://github.com/dask/dask-image/issues/335#issuecomment-1664739810) and recommended [here](https://github.com/dask/dask/issues/10139) still works.
